pkg/agent/openflow/multicluster.go

```diff
@@ -0,0 +1,233 @@
+			MatchDstIPNet(peerServiceCIDR).
+			Action().SetSrcMAC(localGatewayMAC).                 // Rewrite src MAC to local gateway MAC.
+			Action().SetDstMAC(GlobalVirtualMACForMulticluster). // Rewrite dst MAC to virtual Multi-cluster MAC.
+			Action().SetTunnelDst(tunnelPeer).                   // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+		// This generates the flow to forward cross-cluster reply traffic based
+		// on Gateway IP.
+		L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchDstIP(remoteGatewayIP).
+			MatchCTStateRpl(true).
+			MatchCTStateTrk(true).
+			Action().SetSrcMAC(localGatewayMAC).
+			Action().SetDstMAC(GlobalVirtualMACForMulticluster).
+			Action().SetTunnelDst(tunnelPeer). // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+	)
+	return flows
+}
+
+func (f *featureMulticluster) tunnelClassifierFlow(tunnelOFPort uint32) binding.Flow {
+	return ClassifierTable.ofTable.BuildFlow(priorityHigh).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchInPort(tunnelOFPort).
+		MatchDstMAC(GlobalVirtualMACForMulticluster).
+		Action().LoadRegMark(FromTunnelRegMark).
+		Action().LoadRegMark(RewriteMACRegMark).
+		Action().GotoStage(stageConntrackState).
+		Done()
+}
+
+func (f *featureMulticluster) outputFlow(tunnelOFPort uint32) binding.Flow {
+	return L2ForwardingOutTable.ofTable.BuildFlow(priorityHigh).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchRegMark(RewriteMACRegMark).
+		MatchRegMark(OutputToTunnelRegMark).
+		MatchInPort(tunnelOFPort).
+		Action().OutputInPort().
+		Done()
+}
+
+// snatConntrackFlows generates the flows to match the first packet of a cross-cluster connection.
+// SNAT related flows are only installed on Gateway Node.
+func (f *featureMulticluster) snatConntrackFlows(serviceCIDR net.IPNet, localGatewayIP net.IP) []binding.Flow {
+	var flows []binding.Flow
+	ipProtocol := getIPProtocol(localGatewayIP)
+
+	flows = append(flows,
+		// This generates the flow to match the first packet of a Service connection sourced from a local Pod or local
+		// Node, and the Endpoint is a Service ClusterIP on remote cluster. Note that, DNAT has been performed,
+		// and SNAT should also be performed before sending the packets to the remote cluster.
+		// This flow is used to mark the connection requiring SNAT by loading ConnSNATCTMark in DNAT CT zone.
+		ServiceMarkTable.ofTable.BuildFlow(priorityHigh).
+			Cookie(f.cookieAllocator.Request(f.category).Raw()).
+			MatchProtocol(ipProtocol).
+			MatchCTStateNew(true).
+			MatchCTStateTrk(true).
+			MatchCTMark(ServiceCTMark).
+			MatchDstIPNet(serviceCIDR).
+			Action().CT(true, ServiceMarkTable.GetNext(), f.dnatCtZones[ipProtocol], nil).
+			LoadToCtMark(ConnSNATCTMark).
```

Comment:
I don't quite get how the 3 flows work together, have 3 questions:
1. The 1st flow identifies service traffic whose backend is a service IP in another cluster and loads ConnSNATCTMark, but the 2nd flow has same checks and never use the mark, is there redundancy between them and why loading mark is required?
2. The 2nd flow and the 3rd flow seems covering all traffic to remote service ip regardless of traffic source, then why we need two flows for DNATed traffic and non-DNATed traffic separately?
3. The 3rd flow loads ConnSNATOnlyCTMark but it's only used in a table before this flow, is it for subsequent packets? If yes, why is it needed? If the first packet is not DNATed, how could subsequent packets be DNATed even without the mark?

Reply:
I think the reason why `ConnSNATOnlyCTMark` and `conntrackFlows` are required is that subsequent request packets of a SNATed connection is SNATed by "UNSNAT" table unexpectedly, leading to an unknown but established connection shown in DNAT zone. If we SNAT sebsequent request packets in the same place as the first packet, the problem is resolved. And actually it makes more sense to enforce NAT actions in same order for packets in same direction.
I created #3858 to change the existing code to make UNSNAT table handle reply packets only by matching the destination IP. I think if multicluster SNAT do the same, it should add a flow with similar check to UNSNAT table, then we don't need `ConnSNATOnlyCTMark` and `conntrackFlows`.